### DISCUSSION TITLE
Is there a way in RTK query to "land" updates from different API calls

### DISCUSSION THREAD

**Discussion by steinarb** [reduxjs/redux-toolkit#4772] [General]
One thing I do in redux-and-saga is to piggyback updated values on an HTTP POST to update a value.

It's not REST, strictly speaking. I.e. I don't return the object I'm updating.

Instead I usually return the parent of what I'm updating and ["land" it in the same reducer as the data is fetched into with the regular HTTP GET](https://github.com/steinarb/handlereg/blob/9c53e6afe37c88a7cc19e257d789dc23e5a1fb20/handlereg.web.frontend/src/main/frontend/src/reducers/butikkerReducer.js#L11).  Then everything is automatically updated. E.g.
```
const butikkerReducer = createReducer(defaultState, builder => {
    builder
        .addCase(BUTIKKER_MOTTA, (state, action) => action.payload)
        .addCase(NYBUTIKK_LAGRET, (state, action) => action.payload)
        .addCase(BUTIKK_LAGRET, (state, action) => action.payload);
});
```
(BUTIKKER_MOTTA is the result of an HTTP GET for a list of stores, NYBUTIKK_LAGRET is the result of an HTTP POST to create a new store, and contains the complete list of stores, and BUTIKK_LAGRET is the result of an HTTP POST to update a store).

The resulting behaviour is quite nice, and, I think, feels quite naturatual.

Anyway: this is simple to do when the reducer results in a simple array (or object) on the top of the redux state.

Not so easy to put something into the reducer of an RTK query (and one probably shouldn't? Probably no end of contracts broken by doing this?)

But the question remains: is there a simple way in RTK query to achieve the same effect?

Failing that: is there a way for an update action to trigger a reload of a different query? (less network efficient since the data has already arrived in the react app, but the visual effect should be the same)

Thanks!

**Comment by steinarb** [comment]
Here is a similar question on stackoverflow: https://stackoverflow.com/questions/78994431/how-to-trigger-data-fetching-from-an-action-with-rtk-query

I have to admit none of the answers looked very attractive (for the same reasons the person posing the question didn't like them, probably...? "useEffect"...? [The existence of "useEffect" is one reason I'm sticking with redux....](https://steinar.bang.priv.no/2022/06/28/yep-im-still-using-redux/)).

There are actually two things I find myself missing:

1. The ability to land multiple HTTP calls' result in the same reducer
2. The ability for the successful completion of an HTTP call to trigger a new one (because I need data from the first call to have the information needed to do the second call) (probably the thing the stackoverflow OP needs...?)

**Comment by steinarb** [comment]
This has promise, maybe...? https://www.reddit.com/r/reactjs/comments/ruekqt/comment/hqylrjw/

The lazy query option looks complicated...

The conditional fetching sounds promising, but I couldn't figure out where conditional fetching was, in the example linked to...=

**Comment by phryneas** [comment]
I gotta be honest, I don't understand the question: 
do you want to store the result of multiple independent http requests in a slice outside the RTK Query slice?

You can use `extraReducers` and matchers for that.

**Reply by steinarb** [reply]
Well there are actually two questions mixed into one (because they affect the same place in the redux/axios/saga code I'm replacing):

1. One is what you say: store the result of multiple requests in the same place. I will investigate extraReducers for this at some point (but not right now, since I then I also will have to figure out how to initially trigger the requests)
2. Second is simpler: I want the completion of one HTTP request to trigger a new HTTP request (and I'd very much prefer an answer that didn't involve useEffect... :smile: )

**Reply by steinarb** [reply]
To clarify the second question: what I do a lot, is daisy chain sagas doing HTTP requests (when I have successful data from one completed HTTP request, I use that data to compose the redux action triggering a new HTTP request.

Here is an example of the code I'm currently replacing: https://github.com/steinarb/handlereg/blob/9c53e6afe37c88a7cc19e257d789dc23e5a1fb20/handlereg.web.frontend/src/main/frontend/src/sagas/oversiktSaga.js#L21

**Reply by phryneas** [reply]
I'm sure we have this in our docs but I'm at mobile right now so I'm just quoting the first thing for dependent queries that Google gave me: https://github.com/reduxjs/redux-toolkit/discussions/1683#discussioncomment-1583546

If you don't care about the individual results and this will always be "one action together", you can also write a queryFn that calls baseQuery three times, combining three request into just one endpoint.

See https://redux-toolkit.js.org/rtk-query/usage/customizing-queries#customizing-queries-with-queryfn

**Reply by phryneas** [reply]
For the extra reducers thing, see e.g. this example https://redux-toolkit.js.org/rtk-query/usage/examples#using-extrareducers

**Reply by steinarb** [reply]
Ah![ For the first one: nice!](https://github.com/reduxjs/redux-toolkit/discussions/1683#discussioncomment-1583546) I was planning to pick out the isLoading from the first call and do the call to the second one inside an if-statment (not sure if calling hooks inside if-statements are allowed? Probably not...?).  

But dependent queries looks way simpler. Will try it now. (may even be way simpler than the code it is replacing...?)

For the extra reducers thing: thanks! will look at it later! (first cut will be an as-simple-as-possible replacement of redux/axios/saga with RTK query (even if that will not give the exact same behaviour))

**Reply by steinarb** [reply]
Not able to get dependent queries to work: https://gist.github.com/steinarb/2405bf08514c8cb0304193112203716f#file-home-js-L18

The HTTP call from the useGetHandlingerQuery() hook is never called.

Is the destructuring I do, the problem? 

Is the renaming of the values the problem?

Have I misspelled the argument? I've copy-pasted isFulfilled from [the example in the comment](https://github.com/reduxjs/redux-toolkit/discussions/1683#discussioncomment-1583546).

**Reply by markerikson** [reply]
@steinarb there is no `isFulfilled` field in results. Looks like that comment was done while RTKQ was still in alpha, and we presumably removed that field name early on.  (Note that using TS would have caught this error.)

The actual field names are here:

- https://redux.js.org/tutorials/essentials/part-7-rtk-query-basics#query-hook-result-objects

**Reply by steinarb** [reply]
Thanks! Using isSuccess instead of isFulfilled worked: https://gist.github.com/steinarb/1406fcb0c18c6ef673df4e39eebadadb#file-home-js-L18

**Comment by steinarb** [comment]
> But the question remains: is there a simple way in RTK query to achieve the same effect?

Answering my own question here: the RTK query way to achieve the same effect is to use a mutation and then use either onQueryStarted or onCacheEntryAdded to either route the mutation POST's return value into a reducer or trigger a reload of the data using the GET query (I haven't figured out quite how and what yet. Need to figure out what "caches" and "tags" means in this context...?).

(I hadn't gotten to mutations when I started this thread. I started in one end replacing the initial GET operations that populate the app, and now I've come to the first operation that makes a change)

Side note to the RTK people: the JavaScript tab on the examples in the RTK documentation is currently non-functional. I don't know if this is intentional or not, but it makes it harder for people like me (JavaScript users) to use the docs. I would be very happy is this is fixed. :pray:

**Reply by markerikson** [reply]
@steinarb :

> use either onQueryStarted or onCacheEntryAdded to either route the mutation POST's return value into a reducer

Note that you can listen to query-related actions in reducers without having to re-dispatch them:

- https://redux.js.org/tutorials/essentials/part-8-rtk-query-advanced#updating-toast-display

> the JavaScript tab on the examples in the RTK documentation is currently non-functional

Yeah, known issue - we updated a lot of our build tooling for React 19 compat recently, but it also ended up breaking the client-side interactivity on the website. I've got a draft PR up to fix that and should hopefully have it addressed soon.

**Comment by steinarb** [comment]
Note that I haven't been able to use the above information to handle routing of return from updates yet, but this crateApi created matchers solved a different issue for me: https://github.com/reduxjs/redux-toolkit/issues/4778#issuecomment-2549845368

One thing I'm wondering: If a reducer uses a createApi created matcher to store the results of a query or a mutate, is useSelector() of that reducer's value in a component enough to trigger reloads of that value? Or do I still need to use the useBlahBlahQuery() hook to make the HTTP loading go?

**Reply by markerikson** [reply]
@steinarb `useSelector` triggers re-renders based on whatever selector result you return - doesn't matter how the reducer determined it needed to update, it'll re-render if that value _does_ update.

**Reply by steinarb** [reply]
Yeah, that's what I thought! (but it's always better to ask than to assume)

(I'm still working the issue of utilizing the return from a mutation instead of doing another roundtrip (including an extra database operation). I looked at extrareducers but couldn't figure out how they fit in...? With matchers from the api.endpoints I am able to land successful returns in regular reducers, so I could land both the query and the mutation result in the same place and then useSelector() that into the component. But then I still have the issue of how to trigger the HTTP operations)

**Reply by EskiMojo14** [reply]
useSelector will not manage the actual process of fetching, you still need to use useQuery for that.

**Reply by steinarb** [reply]
Yup, I got that part.

**Reply by steinarb** [reply]
I think what I'm looking for is something like the `lifecycleApi.dispatch(apiSlice.util.updateQueryData())` seen here: https://redux.js.org/tutorials/essentials/part-8-rtk-query-advanced#implementing-optimistic-updates

Except I want to use the return data of the post, after the post returns, and use that to replace the entire get results (instead of pathcing the get results when the post is done).

Am I on the track here? Or barking up the wrong tree?

Other than saving roundtrips by using mutation POST results, the conversion from axios/saga to RTK query/RTK listeners, is done: sagas and axios are both gone, reducer list and has been reduced quite a bit, and the redux actions list has been seriously reduced.

cloc results before conversion:
```
      91 text files.
      89 unique files.                              
       5 files ignored.

github.com/AlDanial/cloc v 1.96  T=0.03 s (3148.6 files/s, 83795.2 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
JavaScript                      83            236             39           2007
JSON                             3              0              0             50
SVG                              1              0              0              7
CSS                              1              0              0              3
-------------------------------------------------------------------------------
SUM:                            88            236             39           2067
-------------------------------------------------------------------------------
```

cloc results after conversion:
```
      60 text files.
      58 unique files.                              
       5 files ignored.

github.com/AlDanial/cloc v 1.96  T=0.03 s (2001.6 files/s, 53621.9 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
JavaScript                      52            129             39           1301
JSON                             3              0              0             48
SVG                              1              0              0              7
CSS                              1              0              0              3
-------------------------------------------------------------------------------
SUM:                            57            129             39           1359
-------------------------------------------------------------------------------
```

So... unless I've miscalculated: a 35% reduction of the JavaScript source size...? Not bad, not bad!

Also: looking at the devtools network tab: a lot fewer network calls, with just as good updated data.

**Reply by markerikson** [reply]
@steinarb : yes, that would be a "pessimistic update":

- https://redux-toolkit.js.org/rtk-query/usage/manual-cache-updates#pessimistic-updates

note that while the example snippet in that docs page shows `api.util.updateQueryData`, you probably want to use the newer `upsertQueryEntries` API instead:

- https://redux-toolkit.js.org/rtk-query/api/created-api/api-slice-utils#upsertqueryentries

**Reply by steinarb** [reply]
Here is the API before adding pessimistic updates (this works as expected but has an extra fech of data already received) https://gist.github.com/steinarb/1d6e9ba31a22b811e21211ea503bf43b

I have attempted to do pessimistic updates, and I have added console.log traces and seen that they are called when expected, but the data from the get endpoints aren't replaced: https://gist.github.com/steinarb/0e85a37bc4c5c4e53a37c4b3f9299b31#file-api-js-L33

I think the problem may be in the second argument of upsertQueryEntries https://gist.github.com/steinarb/0e85a37bc4c5c4e53a37c4b3f9299b31#file-api-js-L38

In lack of a better alternative I put the POST argument there, but from the upsertQueryEntries docs it looks like the GET arguments should be there...? https://redux-toolkit.js.org/rtk-query/api/created-api/api-slice-utils#upsertqueryentries

However getButikker has no argument (it returns a list of all stores, which is the same as the postEndrebutikk mutation returns) so I don't have an argument to put there...? 

I tried using null, as the second argument, but that didn't make upsertqueryentries update anything either.

**Reply by steinarb** [reply]
Hm... undefined instead of null, maybe...? https://stackoverflow.com/a/75249499

**Reply by steinarb** [reply]
Nope, undefined had the same effect as null, i.e. no update of the getButikker results.

**Reply by steinarb** [reply]
Tried replacing upsertQueryEntries with updateQueryData like in the stackoverflow answer, but that didn't make the getButikker endpoint data being updated, either:
```
        postEndrebutikk: builder.mutation({
            query: (body) => ({ url: '/endrebutikk', method: 'POST', body }),
            async onQueryStarted(body, { dispatch, queryFulfilled }) {
                try {
                    const { data: postEndrebutikkResult } = await queryFulfilled;
                    dispatch(api.util.updateQueryData('getButikker', undefined, postEndrebutikkResult));
                } catch {}
            },
        }),
```

**Reply by steinarb** [reply]
If it had worked I probably should have seen a redux action in redux devtool...? 

And I don't:
![image](https://github.com/user-attachments/assets/d70d5d0e-2731-4fdf-8233-0ce28714aaa1)

But console.log outputs a sensible value for postEndrebutikkResult and a sensible-looking value for dispatch at the expected time ...? 
```
        postEndrebutikk: builder.mutation({
            query: (body) => ({ url: '/endrebutikk', method: 'POST', body }),
            async onQueryStarted(body, { dispatch, queryFulfilled }) {
                try {
                    const { data: postEndrebutikkResult } = await queryFulfilled;
                    console.log('onQueryStarted(1) postEndrebutikkResult');
                    console.log(postEndrebutikkResult);
                    console.log(dispatch);
                    dispatch(api.util.updateQueryData('getButikker', undefined, postEndrebutikkResult));
                } catch {}
            },
        }),
```

**Reply by steinarb** [reply]
There was an exception thrown from dispatch()

I did this:
```
        postEndrebutikk: builder.mutation({
            query: (body) => ({ url: '/endrebutikk', method: 'POST', body }),
            async onQueryStarted(body, { dispatch, queryFulfilled }) {
                try {
                    const { data: postEndrebutikkResult } = await queryFulfilled;
                    console.log('onQueryStarted(1) postEndrebutikkResult');
                    console.log(postEndrebutikkResult);
                    console.log(dispatch);
                    const updateResult = dispatch(api.util.updateQueryData('getButikker', undefined, postEndrebutikkResult));
                    console.log('onQueryStarted(2) updateResult');
                    console.log(updateResult);
                } catch(e) {
                    console.log('onQueryStarted(3) catch');
                    console.log(e);
                }
            },
        }),
```
and got the following output:
````
onQueryStarted(1) postEndrebutikkResult
api.js:39 
(138) [{…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, …]
api.js:40 (action, ...args) => dispatch(action, ...args)
api.js:45 onQueryStarted(3) catch
api.js:46 Error: [Immer] minified error nr: 6. Full error at: https://bit.ly/3cXEKWf
    at onQueryStarted (api.js:41:42)
````

This is message number 6, I think...? https://github.com/immerjs/immer/blob/main/src/utils/errors.ts#L20

**Reply by steinarb** [reply]
This worked: https://gist.github.com/steinarb/10761e11558e413bc96c57afee7c7181#file-api-js-L25

Made the third argument a single argument lambda, replacing the argument object with the data returned by the mutation HTTP POST.

I'm going with updateQueryData rather than the more modern upsertQueryEntries, since:

1. I feel I understand better  what updateQueryData does
2. I'm replacing the whole GET result anyway, no need for partial replacement

Thanks! :smile:

**Reply by markerikson** [reply]
@steinarb `upsertQueryEntries` _is_ the better tool, and it does the same thing but faster.

There is no "second argument" - it takes an array of one or more entries:

```ts
dispatch(
  api.util.upsertQueryEntries([
    {
      endpointName: "getTodos",
      arg: undefined,
      value: [ /* some todo objects here*/ ]
    },
    {
      endpointName: "getTodo",
      arg: 1,
      value: {id: 1, completed: false}
    },
    // as many more entries as you want, for whatever endpoints you want
  ])
)
```

**Reply by steinarb** [reply]
"same thing but faster" sold me!

I will try to use it (and use it right, this time...).

**Reply by steinarb** [reply]
Hm... I don't know where/when I got it wrong but...
```
TypeError: api.util.upsertQueryEntries is not a function
    at onQueryStarted (api.js:28:39)
```

Should be upsertQueryData... sorry about that!

Edit: with @reduxjs/toolkit 2.5.0 then upsertQueryEntries is in place and seems to work quite well!

**Reply by steinarb** [reply]
Something happens with upsertQueryEntries, in the redux devtool.

But the data hooks like useGetOversiktQuery() returns, isn't updated.

Here is what I do: https://gist.github.com/steinarb/90d4fb317643c9a9441924d4a928a6f1#file-api-js-L23
Here is where it is used and has an old value: https://gist.github.com/steinarb/8d4719556b06099637f697829bc29420#file-kvittering-js-L7

But the values displayed are the old values here (screenshot from redux devtool), i.e. the current values in api->queries->getOversikt->data (the component Kvittering is the yellow rectangle, in the list the number 666 is three transactions from the end, also highlighted in the state tree (with the most recent redux action selected as current)):
![image](https://github.com/user-attachments/assets/d950ca6c-a95c-49cb-a8b6-360ac6d9399c)

**Reply by markerikson** [reply]
Sorry, not sure what you're describing as the issue here.  That `666` value is in the store, and it's also shown in the UI - what are you expecting instead?

**Reply by steinarb** [reply]
I'm expecting 888 as in the last entry of the table. There are two entries that haven't been updated: 777 and 888 in the table.

There is a redux action of type api/queries/cacheEntriesUpserted that changes the values of store.api.queries.getOversikt(undefined).status and store.api.queries.getOversikt(undefined).startedTimeStamp.

But store.api.queries.getOversikt(undefined).data is not updated with the new value (the value returned by the mutation).

**Reply by markerikson** [reply]
I think there's something I'm still not understanding about what your app is supposed to do and what you're expecting. Can you put together a minimal repro with this not working right?

**Reply by steinarb** [reply]
I was expecting (or hoping) store.api.queries.getOversikt(undefined).data to be updated with the data in payload.0.value in the api/queries/cacheEntriesUpserted action.

**Reply by markerikson** [reply]
That's what it should do, yes. But I don't know what that data _was_, because I can't see the rest of the action history, and I can't try out your app myself :)

If you could [record a Replay](https://replay.io/record-bugs) and share it with me, I could at least see what the app is supposed to be doing and what actually happens, but right now all I'm getting from this description is "the data I expected didn't show up".  I don't have enough info to understand the expected sequence here, so I can't tell you whether this is a bug in the library (which tbh is unlikely), a misunderstanding of how `upsertQueryEntries` works, or a timing issue.

**Reply by steinarb** [reply]
Hm... I could push the code of the change would that work?
_Note_: Branch isn't currently pushed (I'm waiting for a conformation before doing so)
 (it would require java JVM >=17, maven and apache karaf to run)

Then 
```
 git clone https://github.com/steinarb/handlereg.git
 cd handlereg
 mvn install
```
and then from the karaf command line
```
feature:repo-add mvn:no.priv.bang.handlereg/karaf/LATEST/xml/features
feature:install handlereg-with-derby
bundle:watch *
```

After this the app will be running on http://localhost:8181/handlereg/
Login with username/password: jad/1ad

(The simplest installation of apache karaf is to download the karaf binary zip, unpack it somewhere and start it with JAVA_HOME referring to a JVM installation https://karaf.apache.org/manual/latest/quick-start.html )

**Comment by steinarb** [comment]
Hm.... I'm running @reduxjs/toolkit 2.2.3 from April 2024 and react 18.2.0 from September (what?) 2022.

Some upgrades are probably in order, but what will break in redux-first-history...? (which is kinda fragile).

Not sure where to start upgrading, but I will try waiting with the new react major version and just try bumping @reduxjs/toolkit first (and if that breaks, bite the bullet and upgrade everything).

**Reply by steinarb** [reply]
Will have to look into [merge callback](https://redux-toolkit.js.org/rtk-query/api/createApi#merge) on the endpoint merged with, maybe...? 

https://redux-toolkit.js.org/rtk-query/api/created-api/api-slice-utils#patchquerydata

**Comment by steinarb** [comment]
@markerikson Just a clarification/explanation of what I'm currently doing and trying to do: :smile: 

1. I am transforming a small react/redux webapp from axios/saga to rtk query/rtk listener as a prototype for doing all of them
2. I have transformed the webapp piece by piece by starting in one saga and start switching it to RTK-query, looking at googled-up examples, rather than reading the docs and doing a thought-through transformation (because this is what works for me)
3. The transformation is complete in that it is working, with all traces of saga removed
4. But the transformed webapp currently uses providesTags on queries and invalidatesTags on mutations to reload data after mutations, and this is slightly inefficient
5. All POST endpoints are constructed so that they return updated data (I do a new JDBC connection and a new select before returning the data), and these returned data are currently thrown on the floor
6. So what i have been doing in this thread (and so far unsuccessfully) is to try to replace the query data with data returned by the mutation (total replacement, no patching, and pessimistic since I'm using the returned data)
7. Replacing query data with the results of the mutations, is the last piece I want to do before I call complete on the transformation, and start summarizing up what I've done and create a plan for the rest of the apps

As for what I am expecting of help by posting here, that's easy to answer: I am expecting nothing. :smile:

It's just that when I am working on stuff I tend to mutter to myself when I am working, in whatever forum is available, because formulating things as questions and problems as text helps me fix the problem. I often see the solution as I'm typing a question, and then I end up sending the post anyway (since I've spent so much time typing it).

I am aware (because I have been told so earlier in other places), that is is something people may find annoying, so if it is, I will take my muttering to myself elsewhere.  (with no offense taken :smile:) 

But please note that at least twice in this particular muttering your comments (and the comments of @EskiMojo14 ) have taken me out of a blind alley and/or restarted the conversion when I was ready to call complete. Thanks for that!

And if I go to muttering to myself I would like to post a final message when I figure out where I went wrong (which may be just forgetting to dot and i or stroke a t somewhere (metaphorically)), because I really hate it when I google up someone who as the exact problem and there is no solution to be found. So I try hard not to be that guy... :smile:

(anyway: unless I hear nothing back I'll go back muttering to myself offline and just post the final solution here)

**Reply by steinarb** [reply]
Just on a side note: I have replacing-queries-with-mutation-returns working with updateQueryData now.

And I am replacing the entire current contents of the query results with the mutation returns anyway, so I'm not sure upsertQueryEntries higher efficiency matters...?

If and if it doesn't matter, I may round up the conversion here. :smile:

**Reply by markerikson** [reply]
@steinarb : if it's just one entry at a time, there won't be a big difference.  But I'm definitely still curious why `upsertQueryEntries` seems to not be working for you.  Loosely put, `upsertQueryData` is a complete thunk cycle that dispatches both `pending` and `fulfilled` actions (and only for _one_ entry at a time), whereas `upsertQueryEntries` is a single sync action that acts as a `fulfilled` actions for a whole batch of entries at once.

**Reply by steinarb** [reply]
The updateQueryData version has been merged to master, but I saved the failing upsertQueryEntries attempt and can push that branch if you're interested...? (it has lots of fixup commits)

**Comment by steinarb** [comment]
For the curious here's the writeup after transforming all of my react/redux webapps from axios/saga to RTK query/RTK listener: https://steinar.bang.priv.no/2025/03/07/convert-react-redux-webapp-from-saga-axios-to-rtk-query-and-rtk-listener/
